control_apps/src/main.rs

```diff
@@ -0,0 +1,409 @@
+        }
+    }
+    fn power_on_self_test(&mut self) {
+        // turn on and test devices to look for errors
+        info!("Starting Power-On Self Test...");
+        // hard-coded wait to check
+        thread::sleep(Duration::from_millis(POST_DELAY_MS));
+        info!("POST complete!");
+        // when POST is complete, transition from idle to safe
+        self.safe()
+    }
+    fn update_pwm(
+        &mut self,
+        gps_altitude: f32, // instantaneous altitude in meters from GPS
+        ballast_mass: f32, // ballast mass remining in kg
+    ) {
+        // execute control algorithm and update vent and dump valve PWM values
+        debug!("Tick!");
+        // check abort criteria
+        abort_if_too_low(self, gps_altitude);
+        abort_if_out_of_ballast(self, ballast_mass);
+        // update error
+        let last_error = self.altitude_error;
+        let error = self.target_altitude - gps_altitude;
+        // change mode if applicable
+        if self.altitude_error.abs() > CTRL_ERROR_DEADZONE {
+            if self.altitude_error > 0.0 {
+                // lower altitude for error to converge to zero
+                self.vent();
+            } else {
+                // raise altitude for error to converge to zero
+                self.dump();
+            }
+        } else {
+            // if in dead zone, do nothing
+            self.idle();
+        }
+        // run control algorithm with gain switching
+        if self.get_mode() == ControlMode::Vent {
+            let new_pwm = self.eval_pid(error, last_error, self.gains_vent);
+            self.valve_vent.set_pwm(new_pwm.abs());
+        } else if self.get_mode() == ControlMode::Dump {
+            let new_pwm = self.eval_pid(error, last_error, self.gains_dump);
+            self.valve_dump.set_pwm(new_pwm.abs());
+        }
+        // update saved error
+        self.altitude_error = error;
+    }
+    fn eval_pid(&mut self, error: f32, last_error: f32, gains: PIDgains) -> f32 {
+        let control_effort = gains.k_p * error + gains.k_d * (error - last_error);
+        debug!("--> altitude error: {:}", error);
+        return control_effort;
+    }
+}
+
+// -- FUNCTIONS --
+fn abort_if_too_low(ctrl_manager: &mut ControlMngr, altitude: f32) {
+    if altitude <= CTRL_ALTITUDE_FLOOR {
+        // abort if the altitude is at or below the allowable limit
+        warn!(
+            "Altitude is too low! Current Alt: {:}m / Min allowed: {:}m",
+            altitude, CTRL_ALTITUDE_FLOOR
+        );
+        ctrl_manager.abort()
+    }
+    // otherwise carry on
+}
+
+fn abort_if_out_of_ballast(ctrl_manager: &mut ControlMngr, ballast_mass: f32) {
+    if ballast_mass <= 0.0 {
+        // abort if there is no ballast left
+        warn!("Not enough ballast mass! Ballast mass: {:}kg", ballast_mass);
+        ctrl_manager.abort()
+    }
+    // otherwise carry on
+}
+
+// -- TEST --
+fn test_control_mngr(ctrl_manager: &mut ControlMngr, target_altitude: f32) {
```

Comment:
I think I need to chat with you to figure out whats the current state/next steps.  I am a little bit confused on what will be interacting with this manager as it seems to have some state machine portions in it, but is having vent directly called on it.